modules/src/ics24_host/identifier.rs

```diff
@@ -1,73 +1,109 @@
-use serde_derive::{Deserialize, Serialize};
 use std::str::FromStr;
 
+use serde::{Deserialize, Serialize};
+
 use super::error::ValidationError;
 use super::validate::*;
-use crate::ics24_host::error::ValidationKind;
 
 /// This type is subject to future changes.
+///
 /// TODO: ChainId validation is not standardized yet.
-/// `is_epoch_format` will most likely be replaced by validate_chain_id()-style function.
-/// https://github.com/informalsystems/ibc-rs/pull/304#discussion_r503917283.
+///       `is_epoch_format` will most likely be replaced by validate_chain_id()-style function.
+///       See: https://github.com/informalsystems/ibc-rs/pull/304#discussion_r503917283.
+///
 /// Also, contrast with tendermint-rs `ChainId` type.
-#[derive(Clone, Debug)]
-pub struct ChainId(String);
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+#[serde(from = "tendermint::chain::Id", into = "tendermint::chain::Id")]
+pub struct ChainId {
+    id: String,
+    version: u64,
+}
 
 impl ChainId {
-    /// Creates a new ChainId() given a chain name and an epoch number. May fail by returning an
-    /// error if the epoch is 0.
-    /// The returned chainID will have the format: `{chain name}-{epoch number}`.
+    /// Creates a new `ChainId` given a chain name and an epoch number.
+    ///
+    /// The returned `ChainId` will have the format: `{chain name}-{epoch number}`.
     /// ```
     /// use ibc::ics24_host::identifier::ChainId;
-    /// assert!(ChainId::new("chainA", 0).is_err());
-    /// assert!(ChainId::new("chainA", 1).is_ok());
+    ///
     /// let epoch_number = 10;
-    /// let c_res = ChainId::new("chainA", epoch_number);
-    /// assert!(c_res.is_ok());
-    /// c_res.map(|id| {assert_eq!(ChainId::chain_version(id.to_string()), epoch_number)});
+    /// let id = ChainId::new("chainA".to_string(), epoch_number);
+    /// assert_eq!(id.version(), epoch_number);
     /// ```
-    pub fn new(chain_name: &str, chain_epoch_number: u64) -> Result<Self, ValidationError> {
-        ChainId::from_str(format!("{}-{}", chain_name, chain_epoch_number.to_string()).as_str())
+    pub fn new(name: String, version: u64) -> Self {
+        Self {
```

Comment:
Nice improvement! We'll continue iterating on this very soon probably (https://github.com/informalsystems/ibc-rs/issues/405), as upstream stabilizes on the details of this type.